### DISCUSSION TITLE
Native-query result-mapping expansions (legacy)

### DISCUSSION THREAD

**Discussion by sebersole** [hibernate/hibernate-orm#3955] [Design Proposals]
Thinking we may want to drop our legacy support for native-queries supplying "expansions".  E.g.

```
s.createNativeQuery( "select {f.*}, {c.*} from T_FILE f left join T_FILE c on f.id = c.parent where f.parent is null" )
        .addEntity( "f", File.class )
        .addFetch( "c", "f", "children" )
```

And maybe even the legacy result-mapping API altogether.

WDYT?

**Comment by gavinking** [comment]
Fine by me.

**Comment by beikov** [comment]
I'm not attached to this as I have never used this in the past, but if we drop this, the only way to map native query results is to use `@SqlResultSetMapping` which is not bad, but doesn't support mapping collections. I don't know if there is a HBM way to do such mappings or if that supports collections, but considering the lack of alternatives for collections, I think we still need this, unless you would like to drop support for mapping collections from native queries altogether.

**Reply by sebersole** [reply]
Well, that's what this discussion is partly about - gauging whether people use this.

My gut tells me people generally do not.

There is a hbm.xml way to do this, but the API is just a programatic form of that:

```
	<sql-query name="organization">
		<return alias="f" class="File"/>
		<return-join alias="c" property="f.children"/>
		select {f.*}, {c.*} from T_FILE f left join T_FILE c on f.id = c.parent where f.parent is null
	</sql-query>
```

**Reply by sebersole** [reply]
See also https://trello.com/c/KVV4Smqo/142-sql-resultset-mapping-builder

**Reply by gavinking** [reply]
Well in principle we could add collection loading to what we plan to do with #3879.

But my strong suspicion is that very few people use this.

**Comment by yrodiere** [comment]
I've used this in the past for sure. I don't have access to that code anymore, but I'll try to drop here what I do remember...

My domain model was very simple entities, without associations, essentially a table where each row represents a business operation.

I had to write complex queries with CTEs and UNIONs, and that wasn't supported in HQL at that point. So I had to go native. But really, the result of my query was just entities that were otherwise mapped in Hibernate ORM.

So I wrote a complex native query to select the entities, and then I used this `{f.*}` syntax to "project" the rows back to entities. This worked nicely, though I remember having trouble finding documentation and examples. And I wasn't aware of the `addFetch`.

There may be other ways to do that, but I'm not aware of them.
